Add parameter validation for ChaCha/Poly modules

hanno-arm opened the pull request:
__Summary:__ This PR moves the existing parameter validation for the ChaCha/Poly modules to the framework provided in #1643.

__Important:__ This PR is different from parameter validation work in other modules because the ChaCha20, Poly1305 and ChaChaPoly modules _do_ already have _unconditional_ parameter validation, and this PR only moves it into the framework of #1643. In particular, the previously unconditional parameter checks are now conditional on the (by default disabled) `MBEDTLS_CHECK_PARAMS` configuration option, which is strictly speaking an API break, and it need to be discussed whether this is an acceptable change or not. Ping @mpg @sbutcher-arm.

mpg commented:
>  In particular, the previously unconditional parameter checks are now conditional on the (by default disabled) `MBEDTLS_CHECK_PARAMS` configuration option, which is strictly speaking an API break, and it need to be discussed whether this is an acceptable change or not.

I think that's an acceptable change of behaviour, because code that was previously correct and working will still be correct and working. (I someone wants to argue that relying on passing NULL pointers and getting errors back counts as "correct", with I don't think would be reasonable, we can always argue back that they were relying on undocumented behaviour.)

mpg commented on include/mbedtls/chacha20.h:
```diff
@@ -82,14 +82,15 @@ mbedtls_chacha20_context;
  *                  \c mbedtls_chacha20_free().
  *
- * \param ctx       The ChaCha20 context to initialize.
+ * \param ctx       The ChaCha20 context to initialize. Must not be \c NULL.
```
As in the other PRs: let's use full sentences, including pronouns.

hanno-arm replied:
Ok.

mpg commented on include/mbedtls/chacha20.h:
```diff
@@ -150,16 +153,16 @@ int mbedtls_chacha20_starts( mbedtls_chacha20_context* ctx,
  *                  key and nonce.
  *
  * \param ctx       The ChaCha20 context to use for encryption or decryption.
+ *                  Must be initialized.
```
Suggestion: "It must be initialized and bound to a key and nonce."

(And so on for each function if you decide to adopt that suggestion, I won't repeat it for the other functions and modules.)

hanno-arm replied:
Ok.

mpg commented on include/mbedtls/chacha20.h:
```diff
@@ -121,6 +123,7 @@ int mbedtls_chacha20_setkey( mbedtls_chacha20_context *ctx,
  *                  messages encrypted with the same nonce and key.
  *
  * \param ctx       The ChaCha20 context to which the nonce should be bound.
+ *                  Must be initialized.
```
Suggestion: "It must be initialized and bound to a key."

hanno-arm replied:
Ok.

mpg commented on library/chacha20.c:
```diff
@@ -202,10 +207,8 @@ void mbedtls_chacha20_free( mbedtls_chacha20_context *ctx )
-    {
-        return( MBEDTLS_ERR_CHACHA20_BAD_INPUT_DATA );
-    }
+    CHACHA20_VALIDATE_RET( ctx != NULL );
```
Hey, good news: we're getting our code a little bit smaller when `MBEDTLS_CHECK_PARAMS` is disabled :)

hanno-arm replied:
Yippie!

mpg commented on library/chachapoly.c:
```diff
@@ -176,16 +175,10 @@ int mbedtls_chachapoly_update_aad( mbedtls_chachapoly_context *ctx,
+    CHACHAPOLY_VALIDATE_RET( ctx != NULL );
+    CHACHAPOLY_VALIDATE_RET( aad_len == 0 || aad != NULL );
+
+    if( ctx->state != CHACHAPOLY_STATE_AAD )
```
Minor: style: could remove the braces below to save some vertical space, now that this is a simple if.

hanno-arm replied:
Fixed.

mpg commented on library/poly1305.c:
```diff
@@ -276,27 +282,24 @@ static void poly1305_compute_mac( const mbedtls_poly1305_context *ctx,
-    {
-        mbedtls_platform_zeroize( ctx, sizeof( mbedtls_poly1305_context ) );
-    }
+        return;
```
Err, you obviously mean `if( ctx == NULL )` above then.

hanno-arm replied:
Oh my.

hanno-arm replied:
Fixed.

mpg commented on library/chachapoly.c:
```diff
@@ -90,39 +96,35 @@ static int chachapoly_pad_ciphertext( mbedtls_chachapoly_context *ctx )
-        ctx->state          = CHACHAPOLY_STATE_INIT;
-        ctx->mode           = MBEDTLS_CHACHAPOLY_ENCRYPT;
-    }
+        return;
```
You want to swap the `!=` above for a `==` then.

hanno-arm replied:
Yes. I shouldn't start rushing things, apparently.

mpg commented on tests/suites/test_suite_chacha20.function:
```diff
@@ -93,42 +93,37 @@ void chacha20_bad_params()
     mbedtls_chacha20_context ctx;
 
-    mbedtls_chacha20_init( NULL );
-    mbedtls_chacha20_free( NULL );
```
Please keep a test for `free( NULL )` using `TEST_VALID_PARAM()`.

mpg replied:
(Same for the other modules.)

hanno-arm replied:
How do you suggest to test this using `TEST_INVALID_PARAM` if `NULL` is not an invalid parameter to `mbedtls_xxx_free()`?

k-stachowiak replied:
@hanno-arm You may have already noticed, but @mpg was referring to `TEST_VALID_PARAM()` rather than `TEST_INVALID_PARAM()`. This commit https://github.com/ARMmbed/mbedtls/pull/1643/commits/54e7f312d33581a83e4dde3fbf5750ffbf5d322f makes the macro compiled in regardless of the parameter checking, so you may want to rebase this PR to incorporate that change.

hanno-arm replied:
Thanks @k-stachowiak, and sorry @mpg for not reading your suggestion properly - I wasn't aware of the existence of `TEST_VALID_PARAM`.

hanno-arm replied:
Addressed in ab133a7.

mpg replied:
@hanno-arm "I wasn't aware of the existence of TEST_VALID_PARAM." That's because you didn't re-review #1643 recently ;)

hanno-arm commented:
@mpg Thank you for your review, I addressed your feedback apart from the testing of `free()` functions, for which I'm not yet clear on how you suggested it to work.

k-stachowiak commented on include/mbedtls/aes.h:
```diff
@@ -121,14 +121,14 @@ typedef struct mbedtls_aes_xts_context
  *                 the context.
  *
- * \param ctx      The AES context to initialize.
+ * \param ctx      The AES context to initialize. Must not be \c NULL.
```
These additions should be turned into sentences.

hanno-arm replied:
Thanks, I will fix it.

hanno-arm replied:
Oh, that's actually part of the base PR.

mpg commented:
@k-stachowiak I think Hanno has addressed your previous feedback, please review again.

hanno-arm commented on include/mbedtls/poly1305.h:
```diff
@@ -137,12 +139,12 @@ int mbedtls_poly1305_update( mbedtls_poly1305_context *ctx,
  *                  Authentication Code (MAC).
  *
  * \param ctx       The Poly1305 context to use for the Poly1305 operation.
+ *                  Must be initialized.
```
Fixed.

hanno-arm commented on include/mbedtls/poly1305.h:
```diff
@@ -120,13 +121,14 @@ int mbedtls_poly1305_starts( mbedtls_poly1305_context *ctx,
  *
  * \param ctx       The Poly1305 context to use for the Poly1305 operation.
- * \param ilen      The length of the input data (in bytes). Any value is accepted.
+ *                  Must be initialized.
```
Fixed.

hanno-arm commented on include/mbedtls/poly1305.h:
```diff
@@ -102,11 +103,11 @@ void mbedtls_poly1305_free( mbedtls_poly1305_context *ctx );
  *
  * \param ctx       The Poly1305 context to which the key should be bound.
- * \param key       The buffer containing the 256-bit key.
+ *                  Must be initialized.
```
Fixed.

hanno-arm commented on include/mbedtls/poly1305.h:
```diff
@@ -84,14 +84,15 @@ mbedtls_poly1305_context;
  *                  \c mbedtls_poly1305_free().
  *
- * \param ctx       The Poly1305 context to initialize.
+ * \param ctx       The Poly1305 context to initialize. Must not be \c NULL.
```
Fixed.

hanno-arm commented:
@k-stachowiak @mpg @sbutcher-arm Fixed another commit with minor documentation improvements. Please re-review.

sbutcher-arm commented:
PR #1643 has now been merged, so the 'needs:preceding PR' label can be removed.

sbutcher-arm commented:
Ok, responding to the comment from Hanno:
> This PR is different from parameter validation work in other modules because the ChaCha20, Poly1305 and ChaChaPoly modules do already have unconditional parameter validation, and this PR only moves it into the framework of #1643.

I agree with @mpg. We should have consistency across the modules, and the same error handling in each. Anything different is confusing. So I think we need to make the change. However - this needs to be documented in the ChangeLog under Changes.

cc: @Patater, @RonEld

sbutcher-arm commented on library/chachapoly.c:
```diff
@@ -368,9 +358,12 @@ int mbedtls_chachapoly_auth_decrypt( mbedtls_chachapoly_context *ctx,
     unsigned char check_tag[16];
     size_t i;
     int diff;
```
I personally prefer a clear blank line between variable declarations and the first command.

sbutcher-arm commented on library/poly1305.c:
```diff
@@ -331,16 +334,8 @@ int mbedtls_poly1305_update( mbedtls_poly1305_context *ctx,
     size_t remaining = ilen;
     size_t queue_free_len;
     size_t nblocks;
```
I'd prefer a blank line again.

sbutcher-arm commented on library/chacha20.c:
```diff
@@ -259,15 +260,9 @@ int mbedtls_chacha20_update( mbedtls_chacha20_context *ctx,
-        return( MBEDTLS_ERR_CHACHA20_BAD_INPUT_DATA );
-    }
+    CHACHA20_VALIDATE_RET( ctx != NULL );
+    CHACHA20_VALIDATE_RET( size == 0 || input  != NULL );
```
As I've said in other PR's, we should have a test for `size == 0 || input  != NULL`.  I won't bother making the same observation anywhere else in this PR.

sbutcher-arm commented:
Reviewed, approved and passes CI. Labelling as 'ready to merge'.
